### PR TITLE
feat: add Momento as a standard cache and chat message history provider

### PR DESCRIPTION
**Pull request by malandis**
# Add Momento as a standard cache and chat message history provider

This PR adds Momento as a standard caching provider. Implements the interface, adds integration tests, and documentation. We also add Momento as a chat history message provider along with integration tests, and documentation.

[Momento](https://www.gomomento.com/) is a fully serverless cache. Similar to S3 or DynamoDB, it requires zero configuration, infrastructure management, and is instantly available. Users sign up for free and get 50GB of data in/out for free every month.

## Before submitting

✅ We have added documentation, notebooks, and integration tests demonstrating usage.

## Who can review?

Community members can review the PR once tests pass. Tag maintainers/contributors who might be interested:

cc @hwchase17  @dev2049

### CODE REVIEW

File: langchain/cache.py [hunk 1]
```diff
@@ -26,6 +26,39 @@ def _hash(_input: str) -> str:
...
+    Returns:
+        RETURN_VAL_TYPE: A list of generations.
+    """
+    try:
+        results = json.loads(generations_json)
+        return [Generation(**generation_dict) for generation_dict in results]
+    except json.JSONDecodeError:
+        raise ValueError(
+            f"Could not decode json to list of generations: {generations_json}"
+        )
```
Comment (malandis) [inline langchain/cache.py:59]: I extracted these to functions. I was going to refactor the GPTCache plugin to use these, but found I cannot run their integration tests. They fail on my end and are possibly broken.

File: langchain/memory/chat_message_histories/momento.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+                "Could not import momento python package. "
+                "Please install it with `pip install momento`."
+            )
+
+        self.cache_client: CacheClient = cache_client
+        self.cache_name = cache_name
+        self.session_id = session_id
+        self.key_prefix = key_prefix
+        if ttl is not None:
+            self.ttl = CollectionTtl.of(ttl)
```
Comment (dev2049) [inline langchain/memory/chat_message_histories/momento.py:63]: just checking, is this intentionally different from how ttl is stored in cache?

Comment (malandis) [reply langchain/memory/chat_message_histories/momento.py:63]: Good question -- the Momento cache has a separate way of specifying TTL for collections (lists, dictionaries, sets, etc) to distinguish a TTL on creation vs upsert. `CollectionTtl` encapsulates this extra logic.

Comment (dev2049) [reply langchain/memory/chat_message_histories/momento.py:63]: makes sense, thanks!

**Comment by dev2049** [comment]
@malandis made few minor edits (mostly added some helper constructors that handle client creation), let me know if everything looks good to you!

**Comment by malandis** [comment]
> @malandis made few minor edits (mostly added some helper constructors that handle client creation), let me know if everything looks good to you!

I caught a typo (fixed in [9347cd1](https://github.com/hwchase17/langchain/pull/5221/commits/9347cd1d4b5fd10357a5d8442927f5d8c54d0a1b)) but otherwise looks really clean. Thanks for the lookover!

**Pull request merged**